shell/server/hack-session.js

```diff
@@ -34,12 +36,84 @@ var Url = Npm.require("url");
 var ROOT_URL = Url.parse(process.env.ROOT_URL);
 HOSTNAME = ROOT_URL.hostname;
 
-function HackSessionContextImpl(grainId) {
+function SessionContextImpl(grainId, sessionId) {
   this.grainId = grainId;
+  this.sessionId = sessionId;
 }
 
-makeHackSessionContext = function (grainId) {
-  return new Capnp.Capability(new HackSessionContextImpl(grainId), HackSessionContext);
+SessionContextImpl.prototype.offer = function (cap, requiredPermissions) {
+  var self = this;
+  return inMeteor((function () {
+    var castedCap = cap.castAs(SystemPersistent);
+    var save = castedCap.save();
+    var sturdyRef = waitPromise(save).sturdyRef;
+
+    // TODO(someday): This actually needs to be implemented with the help of the supervisor, and we
+    // should be storing a reference to the parent as well as the requiredPermissions
+    ApiTokens.update({_id: hashSturdyRef(sturdyRef)}, {$set: {requiredPermissions: requiredPermissions}});
+    Sessions.update({_id: self.sessionId}, {$set: {
+      powerboxView: {
+        offer: ROOT_URL.protocol + "//" + makeWildcardHost("api") + "#" + sturdyRef
+      }
+    }});
+  }));
+};
+
+Meteor.methods({
+  finishPowerboxRequest: function (webkeyUrl, saveLabel, grainId) {
+    check(webkeyUrl, String);
+    check(grainId, String);
+
+    var parsedWebkey = Url.parse(webkeyUrl.trim());
+    if (parsedWebkey.host !== makeWildcardHost("api")) {
+      console.log(parsedWebkey.hostname, makeWildcardHost("api"));
+      throw new Meteor.Error(500, "Hostname does not match this server. External webkeys are not " +
+        "supported (yet)");
+    }
+
+    var token = parsedWebkey.hash;
+    if (!token) {
+      throw new Meteor.Error(500, "Invalid webkey. You must pass in the full webkey, " +
+        "including domain name and hash fragment");
+    }
+    token = token.slice(1);
+
+    var cap;
+    try {
+      cap = waitPromise(restoreInternal(token,
+                                        Match.Optional({webkey: Match.Optional(Match.Any)}))).cap;
+    } catch (err) {
+      throw new Meteor.Error(500, err.toString());
+    }
+    var castedCap = cap.castAs(SystemPersistent);
+    var save = castedCap.save({grain: {
+      grainId: grainId,
+      saveLabel: {
+        defaultText: saveLabel
+      }
+    }});
+    var sturdyRef = waitPromise(save).sturdyRef;
+    // TODO(someday): Same as with `offer`, this should eventually be part of the supervisor
+    ApiTokens.update({_id: hashSturdyRef(sturdyRef)}, {$set: {parentToken: hashSturdyRef(token)}});
```

Comment:
This is an abuse of the `parentToken` field. Since you've already saved the capability, the entry in `ApiTokens` already has a provider. You should either manually save the token with `parentToken` as the provider, or you should not set `parentToken` and wait until membranes are implemented (and `save()` will appropriately set the provider).